upgrade bundled maven to 3.9.1

mbien opened the pull request:
highlights.
 - uses `compiler:3.10.1` plugin by default -> `release` compiler option and module info files are now supported out of the box
 - https://maven.apache.org/docs/3.9.1/release-notes.html 
 - https://maven.apache.org/docs/3.9.0/release-notes.html
 
![maven-391](https://user-images.githubusercontent.com/114367/226158855-d725d7f2-2090-4a83-9c87-09b6243e274a.png)



tested
- new project creation, build, run, output parsing etc
- local/remote indexing, hints and queries

observations, saw some extra logging during local repo scan, e.g:
```
[maven-local-indexing] INFO org.eclipse.aether.internal.impl.DefaultArtifactResolver - Artifact org.sonatype.nexus.buildsupport:nexus-buildsupport-bouncycastle:pom:2.14.3-02 is present in the local repository, but cached from a remote repository ID that is unavailable in current build context, verifying that is downloadable from [central (https://repo1.maven.org/maven2/, default, releases+snapshots)]
```
about 20 lines total on a 4 GB .m2 folder - so not too bad.

mbien commented on java/maven.indexer/src/org/netbeans/modules/maven/indexer/NexusRepositoryIndexerImpl.java:
```diff
@@ -526,13 +526,13 @@ private void indexLoadedRepo(final RepositoryInfo repo, boolean updateLocal) thr
-                        p.setProperty("User-Agent", "netBeans/" + System.getProperty("netbeans.buildnumber"));
-                        httpwagon.setHttpHeaders(p);
-                    }
+//                    if (wagon instanceof HttpWagon) { //#216401
+//                        HttpWagon httpwagon = (HttpWagon) wagon;
+//                        //#215343
+//                        Properties p = new Properties();
+//                        p.setProperty("User-Agent", "netBeans/" + System.getProperty("netbeans.buildnumber"));
+//                        httpwagon.setHttpHeaders(p);
+//                    }
```
todo: investigate build error
```
/home/mbien/NetBeansProjects/netbeans/java/maven.indexer/src/org/netbeans/modules/maven/indexer/NexusRepositoryIndexerImpl.java:529: error: cannot access AbstractHttpClientWagon
                    if (wagon instanceof HttpWagon) { //#216401
  class file for org.apache.maven.wagon.shared.http.AbstractHttpClientWagon not found
```

some class moved probably. Could be done via reflection since it is just a method and overall not so important.

mbien commented on java/maven/src/org/netbeans/modules/maven/execute/MavenProxySupport.java:
```diff
@@ -935,7 +949,11 @@ public int next() throws XmlPullParserException, IOException {
         private LineAndColumn startPos() {
             int ln = getLineNumber();
             int col = getColumnNumber();
-            col -= (posEnd - posStart);
+            try {
+                col -= (posEndField.getInt(this) - posStartField.getInt(this));
+            } catch (IllegalAccessException ex) {
+                Exceptions.printStackTrace(ex);
+            }
```
this is ugly, those fields are now private since https://github.com/codehaus-plexus/plexus-utils/commit/797cab4d0fb5ce88ef6863cc504f929f6978ff95

mbien replied:
@sdedic question two: how do I trigger this code path to check if this still works?

I think we have two options here for the mid term: find a different solution for this, or talk to upstream and make the two fields accessible again so that we can remove this hack.

sdedic replied:
This whole thing is triggered when:
- maven `settings.xml` specifies a proxy
- the system does not use a proxy (so no http*_proxy env var or reachable PAC file)
- a maven build or priming build is triggered, or other request in 'online' mode.

I will try to create a simple unittest to verify the settings.xml parsing produces the expected results.

mbien commented:
paperwork failed since I forgot a few things

`MavenDependenciesImplementationTest testDuplicatedDependencies` failed too. This is reproducible locally.

all other tests green, removing all-tests label.

mbien commented on java/maven/test/unit/data/projects/dependencies/golden/testDuplicatedDependencies:
Hi @sdedic, got two questions. Is it ok to update this golden file or does it have to be a specific deps tree? The maven update produces a different tree now which would cause the test to fail.

sdedic replied:
Let me investigate this -- the test specifically checks that duplicate dependencies (possibly included elsewhere in the tree) are enumerated: that the API provides a complete tree with just possible cycles removed.

sdedic replied:
Update: it seems that the dependencies for metadata of `com.fasterxml.jackson.core:jackson-databind:jar:2.13.4:compile` (locally present in .m2 repo) fails with
```
org.apache.maven.project.ProjectBuildingException: Some problems were encountered while processing the POMs:
[ERROR] Failed to determine Java version for profile java14 @ line 223, column 14
[ERROR] Failed to determine Java version for profile java17 @ line 274, column 14
```
this error is not propagated out from the `MavenMetadataSource` -> `DefaultLegacyArtifactCollector` -> `DefaultDependencyTreeBuilder` used in the NB implementation.

sdedic replied:
@mbien -- can I add commit(s) to your branch ? I have a fix for `maven.embedder` that fixes the dependencies.

mbien replied:
@sdedic sure feel free to do so. Before merge I squash it and add you as co-author.

sdedic replied:
OK, the commit also reverts the golden file back to master version.

mbien replied:
nice. so this upgrade found a bug by accident?

sdedic commented:
> 
Strange -- since it uses a specific version of the micronaut and/or other dependencies, the set of dependencies should not change ? Or maybe I did not understand ?

sdedic commented:
Added a simple test that asserts that things parsed out of settings.xml using XPP match the expectation and line/column coordinates point to the approriate text content.

mbien commented:
> > since it diffs dependency trees and the test project uses different versions
> 
> Strange -- since it uses a specific version of the micronaut and/or other dependencies, the set of dependencies should not change ? Or maybe I did not understand ?

you fixed that already, I wrote that comment before pinging you, so its already outdated.

sdedic commented:
I'd recommend to add some notes to our release notes file, namely these things from Maven relnotes might affect the ability of NB to parse / work with existing user files (bcs. of existing user plugins):

> The Maven Resolver transport has changed from Wagon to “native HTTP”, see [Resolver Transport guide](https://maven.apache.org/guides/mini/guide-resolver-transport.html).

Low priority, but if the new http transport behaves differently from the old one in subtle details, some network issues may happen.

> Maven 2.x was auto-injecting an ancient version of plexus-utils dependency into the plugin classpath, and Maven 3.x continued doing this to preserve backward compatibility. Starting with Maven 3.9, it does not happen anymore. This change may lead to plugin breakage.

this seems more important, as the 3rd/4rd party plugins may break bcs. of our upgrade.

> Each line in .mvn/maven.config is now interpreted as a single argument. That is, if the file contains multiple arguments, these must now be placed on separate lines, see [MNG-7684](https://issues.apache.org/jira/browse/MNG-7684).

may affect existing project local configurations

sdedic commented on java/maven/build.xml:
```diff
@@ -56,8 +56,8 @@
             <classpath>
-                <pathelement location="./external/asm-7.0.jar"/>
-                <pathelement location="./external/asm-commons-7.0.jar"/>
+                <pathelement location="./external/asm-9.4.jar"/>
```
Question: asm is provided in platform/libs.asm too, but it seems that since its version is driven by jarjar, it's better to have a separate compile-time external here ?

mbien replied:
i am not sure actually. I was just thinking that the best approach with asm is likely to use the same version everywhere if possible (its sometimes shaded). It is a library which is known to be safe to update.

my thinking was: step 1, have everything use the same version, step 2 in future, check if we can remove the duplicates where it makes sense (if not we just leave it).

mbien replied:
@sdedic has now been deduplicated via #5751

mbien commented:
updated the binariesembedded-list

squashed the upgrade commits into one and squashed the test+fix from @sdedic into another commit and moved that commit in front of the upgrade commit because both can be tested independently.

planning to merge it as is without further squashing.

ebarboni commented:
> I forgot updating: https://github.com/apache/netbeans/blob/4c90ff36e0ac1d80b324b76fca0befc13bd750f6/java/maven.embedder/external/binariesembedded-list
> 
> this is called manually from
> 
> https://github.com/apache/netbeans/blob/4c90ff36e0ac1d80b324b76fca0befc13bd750f6/java/maven.embedder/build.xml#L30-L33
> 
> and also from:
> 
> https://github.com/apache/netbeans/blob/4c90ff36e0ac1d80b324b76fca0befc13bd750f6/nbbuild/build.xml#L248-L252
> 
> this runs checksums again over the contents of the zip, additional to the checksum over the zip itself.
> 
> @ebarboni I suppose I have to update this too?

Yes this is needed :D. Was looking at doing a PR for 3.8.8 but see yours for 3.9.1.

I wonder if jdom1.0 references is needed ? I do a test removing it and it compiles ok. (maven dependency try still needed)

mbien commented:
something went wrong during squashing and the golden file wasn't reverted to the old state. have to investigate.

ebarboni commented:
@mbien forget my comment on jdom it's used in maven.osgi for example.  But could be later a libs like many others ?

mbien commented:
all tests green. double checked that the golden file didn't change. merging.

mbien commented:
> I'd recommend to add some notes to our release notes file

@sdedic added the https://github.com/apache/netbeans/labels/Highlight label so that we don't forget

-> https://github.com/apache/netbeans/pull/5679#issuecomment-1479023630
